<p style="text-align:center"> 
<a href="https://skills.network" target="_blank"> 
<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo"> 
</a>
</p>

# <h1 align="center"><font size="7"><strong>Final project</strong></font></h1>
## <h2 align= "center"><font size="6.8">Homes for sale in King County, USA</font></h2>

<h2>Table of contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px"> 
<ul> 
<li><a href="#Instructions">Instructions</a></li> 
<li><a href="#About-the-Dataset">About the Dataset</a></li> 
<li><a href="#Module-1:-Importing-Data-Sets">Part 1: Importing the data </a></li> 
<li><a href="#Module-2:-Data-Wrangling">Part 2: Data Manipulation</a> </li> 
<li><a href="#Module-3:-Exploratory-Data-Analysis">Part 3: Exploratory Data Analysis (EDA)</a></li> 
<li><a href="#Module-4:-Model-Development">Part 4: Model Development</a></li>
<li><a href="#Module-5:-Model-Evaluation-and-Refinement">Part 5: Model Evaluation and Refinement</a></li>
</a></li>
</div>

<hr>

## Instrucciones

In this role, we are tasked with being a data analyst working for a real estate investment trust. The trust wants to begin investing in residential real estate, so we are tasked with determining the market price of a home, given a number of characteristics.

* <strong>Objective</strong>:

``Analyze and predict home prices using attributes or characteristics such as square footage (sqft), number of bedrooms, number of floors, etc.``


<hr>

## About the Dataset

This dataset contains home sales prices in King County, which includes Seattle. It includes homes sold between May 2014 and May 2015. It was extracted from this [source](https://www.kaggle.com/harlfoxem/housesalesprediction?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-wwwcourseraorg-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2022-01-01). It was also slightly modified for the purposes of this exercise.

### Dataset variables

| Variable | Description |
| ------------- | ----------------------------------------------------------------------------------------------------------- |
| id | A rating for a house |
| date | Date the house sold |
| price | Price of the house (this is our goal to predict) |
| bedrooms | Number of bedrooms |
| bathrooms | Number of bathrooms |
| sqft_living | Square footage of the house |
| sqft_lote | Square footage of the lot |
| floors | Total floors (levels) in the house |
| waterfront | House with an ocean view |
| view | Has been viewed |
| condition | How good the overall condition is |
| grade | Overall rating given to the housing unit, according to the King County rating system |
| sqft_above | Square footage of the house excluding the basement |
| sqft_basement | Square footage of the basement |
| yr_built | Year built |
| yr_renovated | Year the house was renovated |
| zipcode | Postal code |
| lat | Latitude coordinate |
| long | Longitude coordinate |
| sqft_living15 | Living room area in 2015 (involves some renovations). This may or may not have affected the lot area. |
| sqft_lot15 | Lot area in 2015 (involves some renovations) |

<hr>

## Setup

For this work, we will use the following Python libraries:

*   [`pandas`](https://pandas.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) to manage data.
*   [`numpy`](https://numpy.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for mathematical operations.
*   [`scikit-learn`](https://scikit-learn.org/stable/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for machine learning-related functions and the machine learning pipeline.
*   [`seaborn`](https://seaborn.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) to visualize the data.
*   [`matplotlib`](https://matplotlib.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for additional plotting tools.


### **We install and import the mentioned libraries**

In [ ]:
# We install the libraries using a magic command
%pip install pandas numpy matplotlib seaborn scikit-learn

In [2]:
# We import the libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler,PolynomialFeatures
from sklearn.linear_model import LinearRegression

import warnings
warnings.filterwarnings('ignore')

# To make visualizations run inside cells
%matplotlib inline

<hr>

## Part 1: Importing datasets

We directly load the dataset (in this case a file with the extension <code>.csv</code> that comes from an external link) into a data frame in our Jupyter notebook and then have it locally.

In [3]:
# CSV file address
file_path = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/FinalModule_Coursera/data/kc_house_data_NaN.csv'

# Load the CSV file into a Pandas dataframe
df = pd.read_csv(file_path)

But before downloading the uncleaned dataset, we check if the dataframe is saved correctly with the <code>head()</code> method to display the first 5 columns of the dataframe.

In [4]:
df.head()

Unnamed: 0          id             date     price  bedrooms  bathrooms  \
0           0  7129300520  20141013T000000  221900.0       3.0       1.00   
1           1  6414100192  20141209T000000  538000.0       3.0       2.25   
2           2  5631500400  20150225T000000  180000.0       2.0       1.00   
3           3  2487200875  20141209T000000  604000.0       4.0       3.00   
4           4  1954400510  20150218T000000  510000.0       3.0       2.00   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1180      5650     1.0           0  ...      7        1180   
1         2570      7242     2.0           0  ...      7        2170   
2          770     10000     1.0           0  ...      6         770   
3         1960      5000     1.0           0  ...      7        1050   
4         1680      8080     1.0           0  ...      8        1680   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 22 columns]

In [5]:
#We download the uncleaned CSV file so we can compare it with future versions.
df.to_csv('../data/raw/data.csv')

### We perform a brief inspection to see how the dataframe turned out.

``Dataset Dimensions``:

In [6]:
print(f'''
Dataset dimensions with its original data:
_ Number of rows: {df.shape[0]}
_ Number of columns: {df.shape[1]}
''')


Dataset dimensions with its original data:
_ Number of rows: 21613
_ Number of columns: 22



``We check the data types of each column in the DataFrame``:

In [7]:
print('Data types per column')
df.dtypes

Data types per column


Unnamed: 0         int64
id                 int64
date              object
price            float64
bedrooms         float64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

``Statistical summary of the DataFrame``:

This allows us to gain a quick and concise overview of the data frame, providing key information about the data distribution, including metrics such as standard deviation, mean, and other descriptive statistics.

In [8]:
df.describe()

Unnamed: 0            id         price      bedrooms     bathrooms  \
count  21613.00000  2.161300e+04  2.161300e+04  21600.000000  21603.000000   
mean   10806.00000  4.580302e+09  5.400881e+05      3.372870      2.115736   
std     6239.28002  2.876566e+09  3.671272e+05      0.926657      0.768996   
min        0.00000  1.000102e+06  7.500000e+04      1.000000      0.500000   
25%     5403.00000  2.123049e+09  3.219500e+05      3.000000      1.750000   
50%    10806.00000  3.904930e+09  4.500000e+05      3.000000      2.250000   
75%    16209.00000  7.308900e+09  6.450000e+05      4.000000      2.500000   
max    21612.00000  9.900000e+09  7.700000e+06     33.000000      8.000000   

        sqft_living      sqft_lot        floors    waterfront          view  \
count  21613.000000  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean    2079.899736  1.510697e+04      1.494309      0.007542      0.234303   
std      918.440897  4.142051e+04      0.539989      0.086517      0.766318   
min      290.000000  5.200000e+02      1.000000      0.000000      0.000000   
25%     1427.000000  5.040000e+03      1.000000      0.000000      0.000000   
50%     1910.000000  7.618000e+03      1.500000      0.000000      0.000000   
75%     2550.000000  1.068800e+04      2.000000      0.000000      0.000000   
max    13540.000000  1.651359e+06      3.500000      1.000000      4.000000   

       ...         grade    sqft_above  sqft_basement      yr_built  \
count  ...  21613.000000  21613.000000   21613.000000  21613.000000   
mean   ...      7.656873   1788.390691     291.509045   1971.005136   
std    ...      1.175459    828.090978     442.575043     29.373411   
min    ...      1.000000    290.000000       0.000000   1900.000000   
25%    ...      7.000000   1190.000000       0.000000   1951.000000   
50%    ...      7.000000   1560.000000       0.000000   1975.000000   
75%    ...      8.000000   2210.000000     560.000000   1997.000000   
max    ...     13.000000   9410.000000    4820.000000   2015.000000   

       yr_renovated       zipcode           lat          long  sqft_living15  \
count  21613.000000  21613.000000  21613.000000  21613.000000   21613.000000   
mean      84.402258  98077.939805     47.560053   -122.213896    1986.552492   
std      401.679240     53.505026      0.138564      0.140828     685.391304   
min        0.000000  98001.000000     47.155900   -122.519000     399.000000   
25%        0.000000  98033.000000     47.471000   -122.328000    1490.000000   
50%        0.000000  98065.000000     47.571800   -122.230000    1840.000000   
75%        0.000000  98118.000000     47.678000   -122.125000    2360.000000   
max     2015.000000  98199.000000     47.777600   -121.315000    6210.000000   

          sqft_lot15  
count   21613.000000  
mean    12768.455652  
std     27304.179631  
min       651.000000  
25%      5100.000000  
50%      7620.000000  
75%     10083.000000  
max    871200.000000  

[8 rows x 21 columns]

<hr>

## Part 2: Data Manipulation

Once we have our data, we can begin processing it.

The first step is to normalize the data; that is, decide whether we need to clean, complete, or format columns and/or values ​​that might affect our analysis.

### Data Format
As we saw with the ``describe()`` method, we have the only column, 'date,' in ``object`` format, that is, as a categorical variable composed of strings.
We must format it to a numeric data type (int or float) so that it doesn't affect the analysis. For dates, we use the `datetime` data type.

In [9]:
# Format the 'date' column to datetime format
df['date'] = pd.to_datetime(df['date'], format='%Y%m%dT%H%M%S')

# Verify that the conversion has been completed
print(f'Formatted date from object to: {df["date"].dtype}')

Formatted date from object to: datetime64[ns]


We've confirmed that the `describe()` method now takes into account the ``date`` column. This method only considers numeric columns (int, float, or datetime).

In [10]:
df.describe()

Unnamed: 0            id                           date         price  \
count  21613.00000  2.161300e+04                          21613  2.161300e+04   
mean   10806.00000  4.580302e+09  2014-10-29 04:38:01.959931648  5.400881e+05   
min        0.00000  1.000102e+06            2014-05-02 00:00:00  7.500000e+04   
25%     5403.00000  2.123049e+09            2014-07-22 00:00:00  3.219500e+05   
50%    10806.00000  3.904930e+09            2014-10-16 00:00:00  4.500000e+05   
75%    16209.00000  7.308900e+09            2015-02-17 00:00:00  6.450000e+05   
max    21612.00000  9.900000e+09            2015-05-27 00:00:00  7.700000e+06   
std     6239.28002  2.876566e+09                            NaN  3.671272e+05   

           bedrooms     bathrooms   sqft_living      sqft_lot        floors  \
count  21600.000000  21603.000000  21613.000000  2.161300e+04  21613.000000   
mean       3.372870      2.115736   2079.899736  1.510697e+04      1.494309   
min        1.000000      0.500000    290.000000  5.200000e+02      1.000000   
25%        3.000000      1.750000   1427.000000  5.040000e+03      1.000000   
50%        3.000000      2.250000   1910.000000  7.618000e+03      1.500000   
75%        4.000000      2.500000   2550.000000  1.068800e+04      2.000000   
max       33.000000      8.000000  13540.000000  1.651359e+06      3.500000   
std        0.926657      0.768996    918.440897  4.142051e+04      0.539989   

         waterfront  ...         grade    sqft_above  sqft_basement  \
count  21613.000000  ...  21613.000000  21613.000000   21613.000000   
mean       0.007542  ...      7.656873   1788.390691     291.509045   
min        0.000000  ...      1.000000    290.000000       0.000000   
25%        0.000000  ...      7.000000   1190.000000       0.000000   
50%        0.000000  ...      7.000000   1560.000000       0.000000   
75%        0.000000  ...      8.000000   2210.000000     560.000000   
max        1.000000  ...     13.000000   9410.000000    4820.000000   
std        0.086517  ...      1.175459    828.090978     442.575043   

           yr_built  yr_renovated       zipcode           lat          long  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean    1971.005136     84.402258  98077.939805     47.560053   -122.213896   
min     1900.000000      0.000000  98001.000000     47.155900   -122.519000   
25%     1951.000000      0.000000  98033.000000     47.471000   -122.328000   
50%     1975.000000      0.000000  98065.000000     47.571800   -122.230000   
75%     1997.000000      0.000000  98118.000000     47.678000   -122.125000   
max     2015.000000   2015.000000  98199.000000     47.777600   -121.315000   
std       29.373411    401.679240     53.505026      0.138564      0.140828   

       sqft_living15     sqft_lot15  
count   21613.000000   21613.000000  
mean     1986.552492   12768.455652  
min       399.000000     651.000000  
25%      1490.000000    5100.000000  
50%      1840.000000    7620.000000  
75%      2360.000000   10083.000000  
max      6210.000000  871200.000000  
std       685.391304   27304.179631  

[8 rows x 22 columns]

### Remove Unnecessary Data
Let's identify and remove unnecessary columns or rows, or at least those that don't directly influence the data in the dataset, such as in this case <code>"id"</code> and <code>"Unnamed: 0"</code> or <code>"date"</code>, which specifies the day the house was sold.

> **Note:**
> Although we could have skipped the first step, it was decided to do it anyway as a practical demonstration of how to proceed in situations where this variable needs to be formatted.

In [11]:
# We remove irrelevant data, in this case the 'id', 'Unnamed: 0', and 'date' columns.
try:
    df.drop(columns='id', inplace=True)
    df.drop(columns='Unnamed: 0', inplace=True)
    df.drop(columns='date', inplace=True)
    print('The columns have been deleted.')
except KeyError:
    print('The columns have already been deleted.')

The columns have been deleted.


In [12]:
# We verify that the columns were deleted correctly
df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0  221900.0       3.0       1.00         1180      5650     1.0           0   
1  538000.0       3.0       2.25         2570      7242     2.0           0   
2  180000.0       2.0       1.00          770     10000     1.0           0   
3  604000.0       4.0       3.00         1960      5000     1.0           0   
4  510000.0       3.0       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

   zipcode      lat     long  sqft_living15  sqft_lot15  
0    98178  47.5112 -122.257           1340        5650  
1    98125  47.7210 -122.319           1690        7639  
2    98028  47.7379 -122.233           2720        8062  
3    98136  47.5208 -122.393           1360        5000  
4    98074  47.6168 -122.045           1800        7503

We generate a statistical summary again, but now without the columns we discarded in the previous step.

In [13]:
df.describe()

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21600.000000  21603.000000  21613.000000  2.161300e+04   
mean   5.400881e+05      3.372870      2.115736   2079.899736  1.510697e+04   
std    3.671272e+05      0.926657      0.768996    918.440897  4.142051e+04   
min    7.500000e+04      1.000000      0.500000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136     84.402258  98077.939805   
std      828.090978     442.575043     29.373411    401.679240     53.505026   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1951.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4820.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000   21613.000000   21613.000000  
mean      47.560053   -122.213896    1986.552492   12768.455652  
std        0.138564      0.140828     685.391304   27304.179631  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471000   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.230000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

### Identifying NaN, Null, or Missing Values
Now we look for missing values. We can see that we have missing values ​​for the <code> bedrooms</code> and <code> bathrooms</code> columns by looking at the `count` row, where we have 21,600 and 21,603 respectively, instead of the total number of rows in the dataset, which is 21,613.

In [14]:
print("Number of NaN values ​​for the bedrooms column:", df['bedrooms'].isnull().sum())
print("Number of NaN values ​​for the bathrooms column:", df['bathrooms'].isnull().sum())

Number of NaN values ​​for the bedrooms column: 13
Number of NaN values ​​for the bathrooms column: 10


Missing values ​​were identified in our dataset. Various statistical strategies exist to address this problem, such as imputation using the mean of the data or the elimination of incomplete records when their value cannot be reasonably estimated. In this case, we chose to replace the missing values ​​using imputation techniques, with the goal of preserving as much information as possible and not reducing the size of the dataset.

In [15]:
# Replace NaN values ​​with the column average
mean = df['bedrooms'].mean()
df['bedrooms'].replace(np.nan, mean, inplace=True)
print('The data was replaced successfully.')

The data was replaced successfully.


We also replace missing values ​​in the ``'bathrooms'`` column with the mean of the ``'bathrooms'`` column itself.

In [16]:
# Replace NaN values ​​with the column average
mean=df['bathrooms'].mean()
df['bathrooms'].replace(np.nan,mean, inplace=True)
print('The data was replaced successfully.')

The data was replaced successfully.


We check again if we have missing values ​​in our data frame

In [17]:
print("Number of NaN values ​​for the bedrooms column:", df['bedrooms'].isnull().sum())
print("Number of NaN values ​​for the bathrooms column:", df['bathrooms'].isnull().sum())

Number of NaN values ​​for the bedrooms column: 0
Number of NaN values ​​for the bathrooms column: 0


### Identifying Outliers
First, an outlier is an observation in a data set that is significantly different from the rest of the data. These values ​​can be the result of measurement errors, incorrectly entered data, extreme phenomena, or legitimate but rare cases. In this case, we have extreme values, but they are clearly real, since when estimating the price of a house, whether it faces the sea or not will be a clear price trigger. We confirm this by observing that it is the column that generates the largest ranges of values.

Therefore, we will consider determining the IQR by taking into account the upper quartile limit of houses with sea views (df['waterfront']).

We proceed to detect outliers in the relevant columns. To do this, we use the ``"Quartile Range"``,which is defined as follows:

<h3 align='center'><code>IQR = Q3 − Q1</code><h3>

Where:
* `IQR` is the quartile range
* `Q3` is the upper quartile
* `Q1` is the lower quartile

``Any value outside this range is considered an outlier.``

In [18]:
# Filter the DataFrame to get only the rows where waterfront = 1
df_waterfront = df[df['waterfront'] == 1]

# Calculate the upper and lower bounds to detect outliers
Q1 = df_waterfront.quantile(0.25)
Q3 = df_waterfront.quantile(0.75)
IQR = Q3 - Q1

# Filter outliers
df_without_atypicals = df_waterfront[~((df_waterfront < (Q1 - 1.5 * IQR)) | (df_waterfront > (Q3 + 1.5 * IQR))).any(axis=1)]

# If you want to keep the original DataFrame without the outliers where waterfront = 1
df_updated = pd.concat([df[df['waterfront'] != 1], df_without_atypicals])

print(f'The original DataFrame had {df.shape[0]} rows and now has {df_updated.shape[0]} rows.')

The original DataFrame had 21613 rows and now has 21558 rows.


We apply the describe method again to see how the descriptive statistics of the dataset turned out.

In [22]:
df_updated.describe()

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.155800e+04  21558.000000  21558.000000  21558.000000  2.155800e+04   
mean   5.369100e+05      3.373451      2.114396   2076.573755  1.503397e+04   
std    3.537667e+05      0.925113      0.766176    910.236773  4.130076e+04   
min    7.500000e+04      1.000000      0.500000    290.000000  5.200000e+02   
25%    3.210000e+05      3.000000      1.750000   1420.500000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.605500e+03   
75%    6.420000e+05      4.000000      2.500000   2550.000000  1.063700e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21558.000000  21558.000000  21558.000000  21558.000000  21558.000000   
mean       1.494155      0.005010      0.226459      3.408943      7.654374   
std        0.539967      0.070604      0.750328      0.650104      1.171038   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21558.000000   21558.000000  21558.000000  21558.000000  21558.000000   
mean    1786.097041     290.476714   1971.046386     83.509741  98077.921143   
std      823.591225     439.837084     29.366964    399.650901     53.506782   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1952.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4130.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21558.000000  21558.000000   21558.000000   21558.000000  
mean      47.560185   -122.213625    1984.830504   12711.132155  
std        0.138595      0.140748     683.522887   27262.865849  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471300   -122.328000    1490.000000    5100.000000  
50%       47.571900   -122.230000    1840.000000    7620.000000  
75%       47.678100   -122.125000    2360.000000   10063.750000  
max       47.777600   -121.315000    6210.000000  871200.000000

We see how the average value (mean) of the price column is no longer ``5.400881e+05`` but ``5.360473e+05``. Since we limited the range in the total data set by dividing it into two groups, taking the group with the largest data range and discarding the outliers from the latter. All of this to have the largest possible data set to be able to train our future model.

In [23]:
# Save the dataset again, but with the normalized data
df_updated.to_csv('../data/processed/data_clean.csv')

<hr>